src/Processors/QueryPlan/PartsSplitter.cpp

```diff
@@ -212,23 +234,50 @@ ASTs buildFilters(const KeyDescription & primary_key, const std::vector<Values>
     /// Produces ASTPtr to predicate (pk_col0, pk_col1, ... , pk_colN) > (value[0], value[1], ... , value[N]), possibly with conversions.
     /// For example, if table PK is (a, toDate(d)), where `a` is UInt32 and `d` is DateTime, and PK columns values are (8192, 19160),
     /// it will build the following predicate: greater(tuple(a, toDate(d)), tuple(8192, cast(19160, 'Date'))).
-    auto lexicographically_greater = [&](const Values & value)
+    auto lexicographically_greater = [&](const Values & values) -> ASTPtr
     {
-        // PK may contain functions of the table columns, so we need the actual PK AST with all expressions it contains.
-        ASTPtr pk_columns_as_tuple = makeASTFunction("tuple", primary_key.expression_list_ast->children);
-
-        ASTPtr value_ast = std::make_shared<ASTExpressionList>();
-        for (size_t i = 0; i < value.size(); ++i)
+        ASTs pks_ast;
+        ASTs values_ast;
+        for (size_t i = 0; i < values.size(); ++i)
         {
-            const auto & types = primary_key.data_types;
-            ASTPtr component_ast = std::make_shared<ASTLiteral>(value[i]);
+            /// NULL is treated as a terminator for > comparison.
+            if (values[i].isNull())
+                break;
+
+            const auto & type = primary_key.data_types.at(i);
+
+            // PK may contain functions of the table columns, so we need the actual PK AST with all expressions it contains.
+            auto pk_ast = primary_key.expression_list_ast->children.at(i);
+
+            // If PK is nullable, prepend a null mask column for > comparison.
+            if (type->isNullable())
+            {
+                pks_ast.push_back(makeASTFunction("isNull", pk_ast));
+                values_ast.push_back(std::make_shared<ASTLiteral>(0));
+            }
+
+            ASTPtr component_ast = std::make_shared<ASTLiteral>(values[i]);
+            auto decayed_type = removeNullable(removeLowCardinality(primary_key.data_types.at(i)));
             // Values of some types (e.g. Date, DateTime) are stored in columns as numbers and we get them as just numbers from the index.
             // So we need an explicit Cast for them.
-            if (isColumnedAsNumber(types.at(i)->getTypeId()) && !isNumber(types.at(i)->getTypeId()))
-                component_ast = makeASTFunction("cast", std::move(component_ast), std::make_shared<ASTLiteral>(types.at(i)->getName()));
-            value_ast->children.push_back(std::move(component_ast));
+            if (isColumnedAsNumber(decayed_type->getTypeId()) && !isNumber(decayed_type->getTypeId()))
+                component_ast = makeASTFunction("cast", std::move(component_ast), std::make_shared<ASTLiteral>(decayed_type->getName()));
+
+            // If PK is nullable, null mask is used for comparison.
+            // Transform the AST into assumeNotNull(pk) so that the result type is not-nullable.
+            if (type->isNullable())
```

Comment:
imo with only a single branch for `type->isNullable()` it would be easier to read